src/UCTNode.cpp

```diff
@@ -150,41 +141,35 @@ void UCTNode::link_nodelist(std::atomic<int> & nodecount,
     }
 
     // sort (this will reverse scores, but linking is backwards too)
-    std::sort(begin(nodelist), end(nodelist));
-
-    // link the nodes together
-    auto childrenadded = 0;
+    std::sort(rbegin(nodelist), rend(nodelist));
 
     LOCK(get_mutex(), lock);
 
     for (const auto& node : nodelist) {
-        auto vtx = new UCTNode(node.second, node.first, init_eval);
-        link_child(vtx);
-        childrenadded++;
+        auto child = new UCTNode(node.second, node.first, init_eval);
+        m_children.push_back(child);
     }
 
-    nodecount += childrenadded;
+    nodecount += m_children.size();
     m_has_children = true;
 }
 
 void UCTNode::kill_superkos(KoState & state) {
-    UCTNode * child = m_firstchild;
 
-    while (child != nullptr) {
-        int move = child->get_move();
+    auto childIter = begin(m_children);
+    while (childIter != end(m_children)) {
+        int move = (*childIter)->get_move();
 
         if (move != FastBoard::PASS) {
             KoState mystate = state;
             mystate.play_move(move);
 
             if (mystate.superko()) {
-                UCTNode * tmp = child->m_nextsibling;
-                delete_child(child);
-                child = tmp;
+                childIter = m_children.erase(childIter);
```

Comment:
Classic iterator invalidation bug here :-)

All existing end() are no longer valid after this point.

Reply:
Erased element not deleted.

Reply:
That too but that's a general issue with this code. It either has to move the nodes into the vector or use unique_ptrs.

Reply:
`std::stable_partition()` could be used to reorder the nodes in the vector:
```c++
    auto iter = std::stable_partition(begin(m_children), end(m_children),
                                      [&state](const auto &child) {
                                          int move = child->get_move();

                                          if (move != FastBoard::PASS) {
                                              KoState mystate = state;
                                              mystate.play_move(move);
                                              return !mystate.superko();
                                          }
                                          return true;
                                      });
```
nodes that need to be removed are in the range `[iter,end(m_children))`.